flow/record/context.py

```diff
@@ -0,0 +1,53 @@
+from __future__ import annotations
+
+import sys
+from contextlib import contextmanager
+from contextvars import ContextVar
+from dataclasses import dataclass
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from collections.abc import Generator
+
+APP_CONTEXT: ContextVar[AppContext] = ContextVar("app_context")
+
+
+def get_app_context() -> AppContext:
+    """Retrieve the application context, creating it if it does not exist.
+
+    Returns:
+        AppContext: The application context.
+    """
+    if (ctx := APP_CONTEXT.get(None)) is None:
+        ctx = AppContext()
+        APP_CONTEXT.set(ctx)
+    return ctx
+
+
+@contextmanager
+def fresh_app_context() -> Generator[AppContext, None, None]:
+    """Create a fresh application context for the duration of the with block."""
+    token = APP_CONTEXT.set(AppContext())
+    try:
+        yield APP_CONTEXT.get()
+    finally:
+        APP_CONTEXT.reset(token)
+
+
+# Use slots=True on dataclass for better performance which requires Python 3.10 or later.
+# This can be removed when we drop support for Python 3.9.
+if sys.version_info >= (3, 10):
+    app_dataclass = dataclass(slots=True)  # novermin
+else:
+    app_dataclass = dataclass
+
+
+@app_dataclass
+class AppContext:
+    """Context for the application, holding metrics like amount of processed records."""
+
+    records_read: int = 0
+    records_matched: int = 0
+    records_excluded: int = 0
```

Comment:
Could the names be simplified to `read`, `matched`, `excluded`, since it's fairly obvious it's to do with records in this context (🥁)?

Reply:
i'll just keep it AppContext.